src/oauth2c_token_cache_gen_server.erl

```diff
@@ -0,0 +1,144 @@
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%% @doc OAuth2 Authentication Token Cache - This gen_server implements a
+%%% simple caching mechanism for authentication tokens.
+%%% @end
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+%%%_* Module declaration ===============================================
+-module(oauth2c_token_cache_gen_server).
+-behaviour(gen_server).
+
+%%%_* Exports ==========================================================
+%%%_ * API -------------------------------------------------------------
+-export([start/0]).
+-export([start/1]).
+-export([stop/0]).
+-export([get/1]).
+-export([insert/2]).
+-export([delete/1]).
+
+%% gen_server
+-export([init/1]).
+-export([terminate/2]).
+-export([handle_call/3]).
+-export([handle_cast/2]).
+-export([handle_info/2]).
+-export([code_change/3]).
+
+%%%_* Includes =========================================================
+%%%_* Code =============================================================
+%%%_ * Types -----------------------------------------------------------
+%%%_ * API -------------------------------------------------------------
+
+start() -> start(#{cache_ttl => 3600000, cache => #{}}).
```

Comment:
Do you have any suggestions of how to do this? According to https://tools.ietf.org/html/rfc6749, there is an optional field called "expires_in" that contains the token's lifetime in seconds which should be possible to use to set the caching time . However, I am not sure if this field is present. Maybe a better option is to not provide a default value at all and force the end user to specify the lifetime of a cached value?

Reply:
I looked into this and it appears that kivra_core sets the "expiry_time" field (see the issue_token method in https://github.com/kivra/oauth2/blob/master/src/oauth2.erl) which we should be able to use to set the cached token's lifetime dynamically. 

 I've realised that a problem (1) with the current approach, i.e. to set a fixed TTL value for all cached tokens, is that if a token saved in the cache expires while it is cached (and thus not valid anymore); then the retrieve_access_token method will still return this invalid token until it is kicked out from the cache after some fixed time (TTL). Another problem (2) that can occur even though the expiry time of the token is known is if it is revoked by the server before its expiration time.

To prevent the problem (1) described above I suggest one of following approaches:
1. We only cache tokens with a known expiration time; thus, the situation where a token expires while its cached will not occur.
2. We use a default value for tokens with unknown expiry time and implement a mechanism for users of oauth2c to manually evict cache entries if they are no longer valid. If a user receives an invalid token from retrieve_access_token, the user can manually delete the token from the cache and call retrieve_access_token a second time.

I am not sure if problem 2 needs to be handled but I guess it could be solved with the same evict mechanism described as a solution to the problem described above.

Reply:
I noticed that the request method automatically requests a new access token in case of a 401 response. This means that there is no need to implement manual removal of cached values, I can just add the following line to the request method:

`
request(Method, Type, Url, Expect, Headers, Body, Options, Client) ->
  case do_request(Method,Type,Url,Expect,Headers,Body,Options,Client) of
    {{_, 401, _, _}, Client2} ->
     oauth2c_token_cache:delete(TOKEN_KEY), <--- NEW LINE
      {ok, _RetrHeaders, Client3} =
        do_retrieve_access_token(Client2, Options),
      do_request(Method,Type,Url,Expect,Headers,Body,Options,Client3);
    Result -> Result
  end.

With this new insight, I suggest the following approach for setting token cache entries TTLs:
1. If the token's expiry time is known, then use this time as the cached token's expiry time.
2. If the tokens expiry time is unknown, then use a default value.

Reply:
I have implemented the changes suggested in the previous comment.